spec/validators/redirect_uri_validator_spec.rb

```diff
@@ -62,13 +61,18 @@
       expect(subject).to be_valid
     end
 
-    it "accepts native redirect uri" do
+    it "accepts nonstandard oob redirect uri" do
       subject.redirect_uri = "urn:ietf:wg:oauth:2.0:oob"
       expect(subject).to be_valid
     end
 
+    it "accepts nonstandard oob:auto redirect uri" do
```

Comment:
this test seems duplicated? At least i don't see any diff from the one above

Reply:
It is within the "force secured uri" context. Want me to remove it?

Reply:
since we support custom scheme, I think we can also deprecate "force secured uri", @felipeelias

Reply:
I found the commit that introduced the `context` in question: 5ce2e5d. To be honest I'm not sure why there is a special `context` when nothing the the setup of the test really changed. 

@montdidier I think you should remove it since the test is duplicated.

> since we support custom scheme, I think we can also deprecate "force secured uri"
@jasl I'm not sure what you mean by that

Reply:
@felipeelias I mean we can deprecate <https://github.com/doorkeeper-gem/doorkeeper/blob/master/lib/doorkeeper/config.rb#L288>

Reply:
may I ask why?

Reply:
- I think `force_ssl_in_redirect_uri` only disable `http`, we will support custom scheme and
 `force_ssl_in_redirect_uri` may become confuse to users
- As a provider, it allows 3rd apps connect to, `http` redirect uri usually for their test or dev usage, so using `http` redirect uri relates to 3rd app, not provider's Rails env,  one example is Gitlab

Reply:
Thanks for the explanation

> I think force_ssl_in_redirect_uri only disable http, we will support custom scheme and
force_ssl_in_redirect_uri may become confuse to users

If I understood correctly, you would like to drop the `force_ssl_in_redirect_uri` option, since `http` can still be registered as a native url?

> As a provider, it allows 3rd apps connect to, http redirect uri usually for their test or dev usage, so using http redirect uri relates to 3rd app, not provider's Rails env

I agree that `redirect_uri` is 3rd party's responsibility, but shouldn't we try to be secure by default but still let developers change it if needed?

Reply:
> If I understood correctly, you would like to drop the force_ssl_in_redirect_uri option, since http can still be registered as a native url?

Yes

> I agree that redirect_uri is 3rd party's responsibility, but shouldn't we try to be secure by default but still let developers change it if needed?

I think there's no site (Gitlab, Github, Twitter, Google, etc.) disable `http` for OAuth redirect URI, if they use `doorkeeper`, the configure must be set to `false`.

Consider you are doing an app with Github sign in, you probably using `http` in dev and test, if Github force you use `https`, it make your work difficulty.

Reply:
Absolutely, but allowing insecure transport must be a conscious decision by the developer. Example:

```ruby
force_ssl_in_redirect_uri { |uri| uri.host != 'localhost' }
```

Maybe one of the args for this block should also include the `application`, so this could be defined on a per application basis. 

```ruby
force_ssl_in_redirect_uri { |context| !context.application.development? }
```

(assuming development is a flag added by the developer)

What do you think?

Reply:
I think it's better than current.

I forgot there's another reason is deprecating `force_ssl_in_redirect_uri` will help to simplify doorkeeper's code (because we only need to validate schemes are valid form)

Combine this reason and your propose, what about this?

```ruby
uri_scheme_checker do |uri|
  if uri.host != "localhost"
    uri.scheme != "http"
  else
    true
  end
end
```

Reply:
Looks complicated to me, it's hard to understand what does this option do..

Reply:
I'm thinking separate two layers to validate redirect URIs, the first layer is to validate it is an URI, and the second layer is validate its pattern, and the second layer is configurable.

for example, we want to strict not to use `http`, so we configure second layer like my previous comment.

I think this more generic and can be use to other purpose, such as scheme & host whitelists

Reply:
I agree w/ @nbulaj, and for me, it's becoming complicated.

Goal of this PR is to improve native app support while keeping backwards-compatibility so let's stick to the topic. If you want to propose a new way of validating URIs, it deserves a new issue.

Reply:
Understood